phases/19-capstone-projects/72-code-exec-metric/code/main.py

```
@@ -0,0 +1,291 @@
+    script = script.replace("__PLACEHOLDER_CODE__", repr(code))
+    script = script.replace("__PLACEHOLDER_ASSERTIONS__", repr(list(assertions)))
+    return script
+
+
+def run_candidate(code: str, assertions: list[str], timeout_s: float = DEFAULT_TIMEOUT_S) -> ExecResult:
+    if not assertions:
+        return ExecResult(0.0, EXIT_ERROR, 0, 0, detail="no assertions provided")
+    timeout_s = min(max(0.1, float(timeout_s)), MAX_TIMEOUT_S)
+    script = _build_runner(code, assertions)
+    try:
+        proc = subprocess.run(
```

Ruff (0.15.14)

[error] `subprocess` call: check for execution of untrusted input

(S603)